### PR TITLE
Draft cpu usage kernelmode

### PR DESCRIPTION
**Pull request by wacuuu**

### CODE REVIEW

File: docs/storage/prometheus.md [hunk 1]
```diff
@@ -31,6 +31,8 @@ Metric name | Type | Description | Unit (where applicable) | -disable_metrics pa
 `container_cpu_system_seconds_total` | Counter | Cumulative system cpu time consumed | seconds | |
 `container_cpu_usage_seconds_total` | Counter | Cumulative cpu time consumed | seconds | |
 `container_cpu_user_seconds_total` | Counter | Cumulative user cpu time consumed | seconds | |
+`container_cpu_userspace_usage_seconds` | Counter | Time spent in userspace per CPU per container | seconds | |
```
Comment (katarzyna-z) [inline docs/storage/prometheus.md]: Maybe shorter names of metrics will be better?
I think about:
- container_cpu_user_seconds
- container_cpu_kernel_seconds

File: info/v1/container.go [hunk 2]
```diff
@@ -296,6 +296,14 @@ type CpuUsage struct {
 	// Time spent in kernel space.
 	// Unit: nanoseconds.
 	System uint64 `json:"system"`
+
+	// CPU time consumed per core in kernel mode
+	// Units: nanoseconds.
+	PercpuUsageInKernelmode []uint64 `json:"percpu_usage_in_kernelmode"`
```
Comment (katarzyna-z) [inline info/v1/container.go]: Consider shorter name of variable, e.g. `PerCpuKernel` and short json key, e.g: "json:percpu_usage_kernel" .
I think that it fits better to the rest of fields in CpuUsage struct, see: [PerCpu](https://github.com/wacuuu/cadvisor/pull/3/files#diff-4e7bb0420ef22ca45ccd056f4687f8feR290)

File: info/v1/container.go [hunk 3]
```diff
@@ -296,6 +296,14 @@ type CpuUsage struct {
...
+
+	// CPU time consumed per core in user mode
+	// Units: nanoseconds.
+	PercpuUsageInUsermode []uint64 `json:"percpu_usage_in_usermode"`
```
Comment (katarzyna-z) [inline info/v1/container.go]: The same as above, consider changing to `PerCpuUser` and "json:"percpu_usage_user".

File: metrics/prometheus.go [hunk 4]
```diff
@@ -186,6 +186,42 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc, includedMetri
...
 				},
+			}, {
+				name:        "container_cpu_userspace_usage_seconds",
+				help:        "Per core time spent in user mode",
+				valueType:   prometheus.CounterValue,
+				extraLabels: []string{"cpu"},
+				getValues: func(s *info.ContainerStats) metricValues {
+					values := make(metricValues, 0, len(s.Cpu.Usage.PercpuUsageInUsermode))
+					for i, value := range s.Cpu.Usage.PercpuUsageInUsermode {
+						if value > 0 {
```
Comment (katarzyna-z) [inline metrics/prometheus.go]: In my opinion 0 value also gives some information and it's worth to have this metric so there is no need to create this case when value is higher than 0 .
Having this if-case requires additional information about this metric implementation and it's not user friendly.

**Review by katarzyna-z — CHANGES_REQUESTED** [review]

**Comment by katarzyna-z** [comment]
tests failed with following error:
```
--- FAIL: TestMorePossibleCPUs (0.00s)
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0

goroutine 36 [running]:
testing.tRunner.func1.1(0xcfb2c0, 0xc0001a6580)
        /usr/local/go/src/testing/testing.go:941 +0x5d0
testing.tRunner.func1(0xc0001a4d80)
        /usr/local/go/src/testing/testing.go:944 +0x600
panic(0xcfb2c0, 0xc0001a6580)
        /usr/local/go/src/runtime/panic.go:973 +0x3e3
github.com/google/cadvisor/container/libcontainer.setCPUStats(0xc000267f18, 0xc000268180, 0x1)
        /home/kujawka/go/src/github.com/google/cadvisor/container/libcontainer/handler.go:811 +0x892
github.com/google/cadvisor/container/libcontainer.TestMorePossibleCPUs(0xc0001a4d80)
        /home/kujawka/go/src/github.com/google/cadvisor/container/libcontainer/handler_test.go:122 +0x1fd
testing.tRunner(0xc0001a4d80, 0xd63bd8)
        /usr/local/go/src/testing/testing.go:992 +0x1ec
created by testing.(*T).Run

```

File: docs/storage/prometheus.md (hunk 1 above)
Comment (wacuuu) [reply docs/storage/prometheus.md]: Did that, but now you get container_cpu_user_seconds and container_cpu_user_seconds_total in one output. In general i think it is ok, but I'm bringing it up for general awareness.

Comment (katarzyna-z) [reply docs/storage/prometheus.md]: I've seen container_cpu_user_seconds_total metric, imho container_cpu_user_seconds is the proper name of metric but we will see what maintainer decides.

**Comment by katarzyna-z** [comment]
Remember to squash commits to single commit with proper description.

**Comment by katarzyna-z** [comment]
please extend Prometheus tests to handle new metrics, extend https://github.com/wacuuu/cadvisor/blob/4375b78314aa851f13ed46da1c97a8ea1389a261/metrics/prometheus_fake.go#L304 and list of expected Prometheus metrics https://github.com/wacuuu/cadvisor/blob/jwalecki/cpu_kernelmode/metrics/testdata/prometheus_metrics

File: info/v2/container.go [hunk 5]
```diff
@@ -342,6 +342,14 @@ type CpuInstUsage struct {
 	// Time spent in kernel space.
 	// Unit: nanocores per second
 	System uint64 `json:"system"`
+
+	// CPU time consumed per core in kernel mode
+	// Units: nanoseconds.
+	PercpuUsageInKernelmode []uint64 `json:"percpu_usage_in_kernelmode"`
```
Comment (Creatone) [inline info/v2/container.go]: Why this is different than `PerCpuKernel` from `info/v1/container.go`?

File: info/v2/container.go [hunk 6]
```diff
@@ -342,6 +342,14 @@ type CpuInstUsage struct {
...
+
+	// CPU time consumed per core in user mode
+	// Units: nanoseconds.
+	PercpuUsageInUsermode []uint64 `json:"percpu_usage_in_usermode"`
```
Comment (Creatone) [inline info/v2/container.go]: Similar to the above.

File: info/v1/container.go (hunk 2 above)
Comment (katarzyna-z) [reply info/v1/container.go]: I still think that it is better to have shorter names and have the same names in `info/v2/container.go`.
Why have you decided to change it?

File: info/v2/container.go [hunk 7]
```diff
@@ -345,11 +345,11 @@ type CpuInstUsage struct {
 
 	// CPU time consumed per core in kernel mode
 	// Units: nanoseconds.
-	PercpuUsageInKernelmode []uint64 `json:"percpu_usage_in_kernelmode"`
+	PerCpuKernel []uint64 `json:"percpu_usage_in_kernelmode"`
```
Comment (katarzyna-z) [inline info/v2/container.go]: please use the same json tag as for API v1.0  - "percpu_usage_kernel"

File: info/v2/container.go [hunk 8]
```diff
@@ -345,11 +345,11 @@ type CpuInstUsage struct {
...
 
 	// CPU time consumed per core in user mode
 	// Units: nanoseconds.
-	PercpuUsageInUsermode []uint64 `json:"percpu_usage_in_usermode"`
+	PerCpuUser []uint64 `json:"percpu_usage_in_usermode"`
```
Comment (katarzyna-z) [inline info/v2/container.go:352]: the same as above, for API v1.0 you have used "percpu_usage_user"

**Review by katarzyna-z — CHANGES_REQUESTED** [review]
please take a look at my comments

File: info/v2/container.go (hunk 7 above)
Comment (katarzyna-z) [reply info/v2/container.go]: this still needs to be changed to adjust to API v1.0

File: info/v2/container.go (hunk 8 above)
Comment (katarzyna-z) [inline info/v2/container.go]: the same as above

File: container/libcontainer/handler.go [hunk 9]
```diff
@@ -803,9 +803,13 @@ func setCPUStats(s *cgroups.Stats, ret *info.ContainerStats, withPerCPU bool) {
 	}
 	numActual = minUint32(numPossible, numActual)
 	ret.Cpu.Usage.PerCpu = make([]uint64, numActual)
+	ret.Cpu.Usage.PerCpuKernel = make([]uint64, numActual)
+	ret.Cpu.Usage.PerCpuUser = make([]uint64, numActual)
 
 	for i := uint32(0); i < numActual; i++ {
 		ret.Cpu.Usage.PerCpu[i] = s.CpuStats.CpuUsage.PercpuUsage[i]
+		ret.Cpu.Usage.PerCpuKernel[i] = s.CpuStats.CpuUsage.PercpuUsageInKernelmode[i]
```
Comment (katarzyna-z) [inline container/libcontainer/handler.go]: lenght of elements in `s.CpuStats.CpuUsage.PercpuUsageInKernelmode` and in `s.CpuStats.CpuUsage.PercpuUsageInUsermode` is not checked, imho it is not safe.

Comment (wacuuu) [reply container/libcontainer/handler.go]: Is it possible though? Both percpu and percpu in user/kernel come from cgroups. percpu from cpuacct.usage_percpu and per mode is computed from cpuacct.usage_all

Comment (katarzyna-z) [reply container/libcontainer/handler.go]: I think that it is worth to add unit test for case when `len(s.CpuStats.CpuUsage.PercpuUsage) !=0` and `len( s.CpuStats.CpuUsage.PercpuUsageInKernelmode) == len(s.CpuStats.CpuUsage.PercpuUsageInUsermode)  == 0`. What do you think?

**Comment by katarzyna-z** [comment]
@Creatone Could you take a look also at this pull request?

**Pull request closed without merging**